### PR TITLE
Julia ize porta

### PR DESCRIPTION
**Pull request by logantm2**
Converted porta from gnu makefile build to cmake for better platform independence. Cmake now compiles most of the sources into a single shared library (`libporta.so`). The executables are then compiled and linked against this shared library, rather than against each individual compiled object file, as before. Julia programs should be able to simply link against the shared library and [use the standard `ccall` interface](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/).

- Added `CMakeLists.txt`
- Removed `win32` and `gnu-make` dirs
- Moved `xporta` main function into a new source file for cleaner dependency hierarchy

### CODE REVIEW

**Comment by logantm2** [comment]
I tried to use the binarybuilder wizard to test this out, but it errors out for me at step 3:
```
                        # Step 3: Build for Linux x86_64 {libc=glibc}

You will now be dropped into the cross-compilation environment.
Please compile the package. Your initial compilation target is x86_64-linux-gnu
The $prefix environment variable contains the target directory.
Once you are done, exit by typing `exit` or `^D`

ArgumentError: '/tmp/jl_QfdkVc/p3BVe2aR/srcdir' exists. `force=true` is required to remove '/tmp/jl_QfdkVc/p3BVe2aR/srcdir' before copying.
Stacktrace:
  [1] checkfor_mv_cp_cptree(src::String, dst::String, txt::String; force::Bool)
    @ Base.Filesystem ./file.jl:312
  [2] cp(src::String, dst::String; force::Bool, follow_symlinks::Bool)
    @ Base.Filesystem ./file.jl:349
  [3] cp
    @ ./file.jl:349 [inlined]
  [4] setup(source::BinaryBuilderBase.SetupSource{FileSource}, target::String, verbose::Bool)
    @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/IKZGk/src/Prefix.jl:309
  [5] setup_workspace(build_path::String, sources::Vector{Any}, target_platform::Platform, host_platform::Platform; verbose::Bool)
    @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/IKZGk/src/Prefix.jl:376
  [6] step34(state::BinaryBuilder.Wizard.WizardState)
    @ BinaryBuilder.Wizard ~/.julia/packages/BinaryBuilder/CKu9k/src/wizard/interactive_build.jl:523
  [7] run_wizard(state::Nothing)
    @ BinaryBuilder.Wizard ~/.julia/packages/BinaryBuilder/CKu9k/src/Wizard.jl:104
  [8] run_wizard()
    @ BinaryBuilder.Wizard ~/.julia/packages/BinaryBuilder/CKu9k/src/Wizard.jl:87
  [9] top-level scope
    @ REPL[6]:1
 [10] eval
    @ ./boot.jl:360 [inlined]
 [11] eval_user_input(ast::Any, backend::REPL.REPLBackend)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
 [12] repl_backend_loop(backend::REPL.REPLBackend)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
 [13] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
 [14] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
 [15] run_repl(repl::REPL.AbstractREPL, consumer::Any)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
 [16] (::Base.var"#878#880"{Bool, Bool, Bool})(REPL::Module)
    @ Base ./client.jl:387
 [17] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
 [18] invokelatest
    @ ./essentials.jl:706 [inlined]
 [19] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
    @ Base ./client.jl:372
 [20] exec_options(opts::Base.JLOptions)
    @ Base ./client.jl:302
 [21] _start()
    @ Base ./client.jl:485

WizardState [step3]
```

Is this my fault or binarybuilder's fault? I don't know enough about julia to be able to tell

**Comment by logantm2** [comment]
Nvm it was my fault, got it going now

File: README.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ A version of the [PORTA](http://porta.zib.de/) software intended for cross-platf
 
 "PORTA is a collection of [C] routines for analyzing polytopes and polyhedra. The polyhedra are either given as the convex hull of a set of points plus (possibly) the convex cone of a set of vectors, or as a system of linear equations and inequalities." -([PORTA](http://porta.zib.de/)).
 
+## Quick Start
+
+TODO: Add compilation instructions
```
Comment (bdoolittle) [inline README.md]: Please add instructions for how to compile porta from source using CMAKE

File: src/main.c [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    char outfname[FILENAME_SIZE];
+    char fname[FILENAME_SIZE];
+    int   poi_file;
+    int   rowl_inar, ierl;
+    int  *indx = (int *)0;
+    int equa_in,ineq_in, ineq_out;
+    FILE *outfp;
+
+
+
```
Comment (bdoolittle) [inline src/main.c]: let's remove the extraneous whitespace

File: src/main.c [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        for (cone = 0; !(porta_list[cone]->sys[dim].num); cone++);
+        conv = ineq - cone;
+        if (!MP_realised) no_denom(dim+1, 0, cone,1);
+        write_poi_file(*argv,outfp,dim,equa,ineq,cone,0,conv,cone);
+    }
+    else
+        msg( "invalid format of command line", "", 0 );
+
+
+
```
Comment (bdoolittle) [inline src/main.c]: let's remove extraneous whitespace

File: src/main.c [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    fclose( logfile );
+    exit(0);
+}
+
+void resubst( RAT *inner, int equa_in, int indx[] )
+{
+    int sysrow,ie,j;
+    RAT *new_sys;
+
+
```
Comment (bdoolittle) [inline src/main.c]: remove extraneous whitespace

**Review by bdoolittle — CHANGES_REQUESTED** [review]
Thanks @logantm2, this is a big help. The big thing to do is to add the cmake compilation instructions to the readme. the other comments are just removing some extra newlines in the main file.

Once that is done,  I'll go  test the compilation process on Mac OS and test the cross compilation with binary builder

File: README.md (hunk 1 above)
Comment (logantm2) [reply README.md]: Ahh I meant to ask where we wanted to put that. Once the installation is set in stone (maybe more hoops to jump through if the wizard doesn't work the first time thru) I'll add that here and remove the current instructions from `INFO`

**Comment by bdoolittle** [comment]
When running `cmake --build .` on macOS, I'm encountering several errors/warnings:

I'm encountering several warnings about using `labs` instead of `abs`. Do you see these as well? or is this environment specific?

I'm encountering a fatal error during `cmake --build .` 

```
/Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:294:9: warning: 
      implicit declaration of function 'resubst' is invalid in C99
      [-Wimplicit-function-declaration]
        resubst(inner,equa_in,indx);
        ^
/Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:317:6: error: 
      conflicting types for 'resubst'
void resubst( RAT *inner, int equa_in, int indx[] )
     ^
/Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:294:9: note: 
      previous implicit declaration is here
        resubst(inner,equa_in,indx);
        ^
1 warning and 1 error generated.
make[2]: *** [CMakeFiles/xporta.dir/src/main.c.o] Error 1
make[1]: *** [CMakeFiles/xporta.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by logantm2** [comment]
> When running `cmake --build .` on macOS, I'm encountering several errors/warnings:
> 
> I'm encountering several warnings about using `labs` instead of `abs`. Do you see these as well? or is this environment specific?
> 
> I'm encountering a fatal error during `cmake --build .`
> 
> ```
> /Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:294:9: warning: 
>       implicit declaration of function 'resubst' is invalid in C99
>       [-Wimplicit-function-declaration]
>         resubst(inner,equa_in,indx);
>         ^
> /Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:317:6: error: 
>       conflicting types for 'resubst'
> void resubst( RAT *inner, int equa_in, int indx[] )
>      ^
> /Users/brian/Documents/Grad School/Research/Code/julia-porta/src/main.c:294:9: note: 
>       previous implicit declaration is here
>         resubst(inner,equa_in,indx);
>         ^
> 1 warning and 1 error generated.
> make[2]: *** [CMakeFiles/xporta.dir/src/main.c.o] Error 1
> make[1]: *** [CMakeFiles/xporta.dir/all] Error 2
> make: *** [all] Error 2
> ```

Yep, I've been seeing those labs warnings. I've been hesitant to fix those since theyre in the OG code, but I can clean them up if we want to.

One of my recent commits fixes that macos error. Curiously gcc didnt complain about those on linux x86, even though it's DEFINITELY an error

File: src/main.c (hunk 2 above)
Comment (logantm2) [reply src/main.c]: done (do you want to reserve the right to resolve these convos? i'll hold off on pressing the big green button)

File: src/main.c (hunk 3 above)
Comment (logantm2) [reply src/main.c]: donezo

File: src/main.c (hunk 4 above)
Comment (logantm2) [reply src/main.c]: donezo

**Comment by bdoolittle** [comment]
Let's keep OG code right now. As I understand, labs vs abs shouldn't affect the working functionality, which is why it's a warning.

 I believe that int truncation issues occur at the fringe of PORTAs working capacity, especially now since computers have become significantly more powerful over the past 20 years. But this stuff is def out of scope for the current work.

Let's leave abs as is

**Comment by logantm2** [comment]
Sounds good.

The BinaryBuilder wizard now works for me on all the default supported platforms. The hinted cmake command works just fine, just gotta add ` julia-porta` to the end of it to point cmake at the source dir.

Another thing: porta originally came with some executable scripts that were just shortcuts for calling the `xporta` and `valid` binaries with some command-line options. I haven't updated the cmake script here to copy those into the install directory. How much do we care about that?

File: README.md (hunk 1 above)
Comment (logantm2) [reply README.md]: odne

**Comment by bdoolittle** [comment]
All platforms build successfully except for MacOS(:x86_64).

I'm experiencing a Build error on `x86_64-apple-darwin14`.

Build script is:
```
cd $WORKSPACE/srcdir
install_license julia-porta/LICENSE
cmake julia-porta
make -j
make install
mkdir -p $libdir
cp libporta.so $libdir/libporta.so
mkdir -p $bindir
cp xporta $bindir/xporta
cp valid $bindir/valid
```

Error has to do with unsupported options for the linker:

```
...
[13:36:10] [ 58%] Linking C shared library libporta.so
[13:36:10] /usr/bin/cmake -E cmake_link_script CMakeFiles/porta.dir/link.txt --verbose=false
[13:36:10] ld: unknown option: -soname
[13:36:10] clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
[13:36:10] make[2]: *** [CMakeFiles/porta.dir/build.make:224: libporta.so] Error 1
[13:36:10] make[2]: Leaving directory '/workspace/srcdir'
[13:36:10] make[1]: Leaving directory '/workspace/srcdir'
[13:36:10] make[1]: *** [CMakeFiles/Makefile2:134: CMakeFiles/porta.dir/all] Error 2
[13:36:10] make: *** [Makefile:150: all] Error 2
```

@logantm2  did you experience this error as well? do you have an idea about how these options can be controlled with `cmake`?

**Comment by logantm2** [comment]
I didn't see that, it might have to do with the `-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}` option in my build script. Also note that if you use `-DCMAKE_INSTALL_PREFIX=$prefix`, cmake will automatically install into the right dir so you don't need to manually move things around

**Review by bdoolittle — APPROVED** [review]

**Comment by bdoolittle** [comment]
I've successfully built the artifacts with BinaryBuilder, I just need to test the ccall interface in julia to make sure that the binaries still work as expected and that the new libporta.so can be used as expected

**Pull request merged**